llvmPackages_{9,12,13,14,15,16,17,18,git}.clang: fix libLTO.dylib path

reckenrode opened the pull request:
## Description of changes

Clang assumes that `libLTO.dylib` is located at `../lib` in the same prefix as `clang`, but that’s not true in nixpkgs. `libLTO.dylib` is actually located at `libllvm^lib/lib.libLTO.dylib`.

This is the first piece to fixing LTO on Darwin. The other is updating ld64, which I will be doing in a separate PR later this week. This PR does not depend on that PR because there is no harm passing the correct path to ld64. It does not perform LTO by default, and trying to use `-flto` even with the correct path remains broken.

Testing was done using my WIP PR for ld64 with clang 16. See https://github.com/NixOS/nixpkgs/issues/19098#issuecomment-2041752524 for output. It can also be validated by setting `NIX_DEBUG=1` when running clang, then confirming that the path passed to `-lto_library` actually exists in the store.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

reckenrode committed:
llvmPackages_{12,13,14,15,16,17,18,git}.clang: fix libLTO.dylib path

Clang assumes that `libLTO.dylib` is located at `../lib` in the same
prefix as `clang`, but that’s not true in nixpkgs. `libLTO.dylib` is
actually located at `libllvm^lib/lib.libLTO.dylib`.

RossComputerGuy commented:
This is an interesting change, I'll give this a look over when I have the time. Right now, my priority is on making all LLVM subpackages common and I am hoping to get that done asap so this PR will probably take a bit for me to look over. I also don't want to have to rebase that PR heh.

alyssais commented on pkgs/development/compilers/llvm/git/clang/default.nix:
```diff
@@ -63,6 +63,10 @@ let
+    '' + lib.optionalString stdenv.hostPlatform.isDarwin ''
+      # Make sure clang passes the correct location of libLTO to ld64
+      substituteInPlace clang/lib/Driver/ToolChains/Darwin.cpp \
+        --replace-fail 'StringRef P = llvm::sys::path::parent_path(D.Dir);' 'StringRef P = "${lib.getLib libllvm}";'
```
Shouldn't this be unconditional, so that Linux users catch it if the replacement starts failing?

reckenrode replied:
Sure, I can make it unconditional.

reckenrode replied:
It’s unconditional now. I also moved it to the right phase (except for clang 16, which was already in `postPatch`).

alyssais commented:
Conflict :(

reckenrode commented:
> Conflict :(

Rebased and dropped clang 9 since it was removed.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/darwin-updates-news/42249/10

Artturin committed:
qt: convert warnings to throw

geekodour committed:
goose: update test skip list with newer tests

fixes #302629

After updating from `3.15.0 -> 3.19.2` build for `goose` was failing.
This was due to the tests failing esp on integration tests which
required the docker darmon to be running.

Tests with similar nature were already skipped, but on the upgraded
version newer integration tests were added. This change adds these
changes to the test skip list aswell.

Skip following tests which require a docker daemon to be running
- TestLockModeAdvisorySession
- TestDialectStore
- TestGoMigrationStats
- TestPostgresSessionLocker

Links:
- Commit adding the new integration tests: https://github.com/pressly/goose/commit/b2c483ada4b3450f4b95396455aaae33c42d2b78

geekodour committed:
fixup! goose: update test skip list with newer tests

r-ryantm committed:
tmuxp: 1.45.0 -> 1.46.0

mweinelt committed:
firefox-unwrapped: 124.0.2 -> 125.0

https://www.mozilla.org/en-US/firefox/125.0/releasenotes/

mweinelt committed:
firefox-bin-unwrapped: 124.0.2 -> 125.0

https://www.mozilla.org/en-US/firefox/125.0/releasenotes/

mweinelt committed:
firefox-esr-115-unwrapped: 115.9.1esr -> 115.10.0esr

https://www.mozilla.org/en-US/firefox/115.10.0/releasenotes/

Artturin committed:
Merge pull request #304082 from Artturin/qtwarningsthrow

qt: convert warnings to throw

mweinelt committed:
Merge pull request #304319 from mweinelt/firefox-125.0

Firefox: 124.0.2 -> 125.0; 115.9.0esr -> 115.10.0esr

RaghavSood committed:
Merge pull request #304240 from geekodour/master

goose: update test skip list with newer tests

reckenrode commented:
@RossComputerGuy I saw that https://github.com/NixOS/nixpkgs/pull/303948 was merged. I’ve rebased this on y

reckenrode commented:
Closing because of the mass ping. I’ll reopen targeting staging with the consolidated clang.
